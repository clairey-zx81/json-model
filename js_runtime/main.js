import { parseArgs } from 'node:util'
import fs from 'node:fs'

// TODO escaping?
const json_path = (path) => '.' + path.join('.')

function report(rep)
{
    if (rep === null || rep.length === 0) return ''

    let reps = []
    for (const [message, path] of rep)
        reps.push(`${json_path(path)}: ${message}`)

    return ' (' + reps.join('; ') + ')'
}

export default function main(checker)
{
    const options = {
      'verbose': { type: 'boolean', short: 'v' },
      'times': { type: 'string', short: 'T' },
      'report': { type: 'boolean', short: 'r' },
    }

    const args = parseArgs({options, allowPositionals: true})

    // args.hasOwnProperty('verbose')
    // args.verbose === undefined
    let times = args.values.times !== undefined ? Number(args.values.times) : 1

    if (isNaN(times))
    {
        console.log(`Unexpected --times: ${args.values.times}`)
        process.exit(1);
    }

    // overhead estimation
    let empty = 0.0
    if (times > 1)
    {
        let n = times
        while (n--)
        {
            const start = performance.now()
            empty += 1000.0 * (performance.now() - start)
        }
        empty /= times
    }

    for (const fname of args.positionals)
    { 
        try {
            fs.readFile(fname, (err, data) => {
                if (err) throw err
                const value = JSON.parse(data)

                // show performance
                // NOTE: jit, calls may take a lot of time at the beginning
                // NOTE: gc, may be irregular…
                if (times > 1)
                {
                    let sum = 0.0
                    let sum2 = 0.0
                    let n = times
                    while (n--) {
                        const start = performance.now()
                        let rep = args.values.report ? [] : null
                        const valid = checker(value, '', rep)
                        rep && (rep.length = 0)
                        let delay = (1000.0 * (performance.now() - start)) - empty  // µs
                        sum += delay
                        sum2 += delay * delay
                    }
                    const avg = sum / times
                    const stdev = Math.sqrt(sum2 / times - avg * avg)
                    console.log(`${fname}: ${args.values.report ? 'rep' : 'nop'}`,
                                `${avg.toFixed(3)} ± ${stdev.toFixed(3)} µs (${empty.toFixed(3)})`)
                }

                // do it anyway for the result
                let rep = args.values.report ? [] : null
                const valid = checker(value, '', rep)
                console.log(`${fname}: ${valid ? 'PASS' : 'FAIL'}${report(rep)}`)
            })
        }
        catch (e) {
            console.log(`${fname}: ERROR (${e})`)
        }
    }
}
