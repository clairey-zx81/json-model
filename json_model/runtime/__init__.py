from .types import (
    Jsonable,
    Path,
    Report,
    PropMap,
    CheckFun,
    RegexFun,
    EntryCheckFun,
)

from .support import (
    # TODO remove _
    UNDEFINED,
    _tname,
    _path,
    _rep,
    Const,
    ConstSet,
    ConstMap,
    is_valid_date,
    is_valid_time,
    is_valid_datetime,
    is_valid_regex,
    is_valid_exreg,
    is_valid_url,
    is_valid_email,
    is_valid_uuid,
    is_valid_json,
    is_unique_array,
    check_constraint,
    value_len,
    main,
)
