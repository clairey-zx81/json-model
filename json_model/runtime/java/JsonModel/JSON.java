package JsonModel;

import java.util.List;
import java.util.Map;
import java.util.Iterator;
import java.io.Reader;

/**
 * Convenient abstraction for JSON libraries such as GSON and Jackson.
 */
public abstract class JSON<T>
{
    public static class Exception extends java.lang.Exception
    {
        public Exception(String s) {
            super(s);
        }
    }

    public abstract T fromJSON(String json) throws Exception;
    public abstract T fromJSON(Reader reader) throws Exception;
    public abstract String toJSON(T o) throws Exception;
    public abstract boolean isNull(T o);
    public abstract boolean isBoolean(T o);
    public abstract boolean asBoolean(T o);
    public abstract boolean isInteger(T o);
    public abstract int asInteger(T o);
    public abstract long asLong(T o);
    public abstract boolean isDouble(T o);
    public abstract double asDouble(T o);

    public boolean isNumber(T o)
    {
        return isInteger(o) || isDouble(o);
    }

    public double asNumber(T o)
    {
        return isDouble(o) ? asDouble(o) : (double) asLong(o);
    }

    public abstract boolean isString(T o);
    public abstract String asString(T o);
    public abstract boolean isArray(T o);
    public abstract int arrayLength(T o);
    public abstract T arrayItem(T o, int index);
    public abstract Iterator<? extends T> arrayIterator(T o);
    public abstract boolean isObject(T o);
    public abstract int objectSize(T o);
    public abstract T objectValue(T o, String prop);
    public abstract Iterator<String> objectIterator(T o);
}
