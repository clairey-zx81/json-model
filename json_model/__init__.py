#
# convenient types, classes and functions to process models
#
from .mtypes import Jsonable
from .runtime.types import Report, EntryCheckFun
from .model import JsonModel
from .script import (
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
)

__ALL__ = [
    Jsonable, Report, EntryCheckFun, JsonModel,
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
]
