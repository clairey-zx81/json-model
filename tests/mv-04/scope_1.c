//
// Generated by JSON Model Compiler version 2.0b0
// see https://github.com/clairey-zx81/json-model
//

// regular expression engine
#include <stddef.h>
#include <cre2.h>

#include <json-model.h>
#define JSON_MODEL_VERSION "2.0b0"

static bool json_model_4(const json_t *val, jm_path_t *path, jm_report_t *rep);
static cre2_regexp_t *_jm_re_0_re2 = NULL;
static int _jm_re_0_nn = 0;
static bool _jm_re_0(const char *s, jm_path_t *path, jm_report_t *rep);
static bool json_model_3(const json_t *val, jm_path_t *path, jm_report_t *rep);
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep);
static cre2_regexp_t *_jm_re_1_re2 = NULL;
static int _jm_re_1_nn = 0;
static bool _jm_re_1(const char *s, jm_path_t *path, jm_report_t *rep);
static bool json_model_5(const json_t *val, jm_path_t *path, jm_report_t *rep);
jm_propmap_t check_model_map_tab[3];
const size_t check_model_map_size = 3;

// check $R (.'$R')
static bool json_model_4(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$R'
    res = json_is_array(val) && json_array_size(val) == 2;
    if (res)
    {
        jm_path_t lpath_0 = (jm_path_t) { NULL, 0, path, NULL };
        // .'$R'.0
        res = json_model_5(json_array_get(val, 0), (path ? &lpath_0 : NULL), rep);
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "unexpected $S [.'$R'.0]", (path ? &lpath_0 : NULL));
        }
        if (res)
        {
            jm_path_t lpath_0 = (jm_path_t) { NULL, 1, path, NULL };
            // .'$R'.1
            res = json_model_5(json_array_get(val, 1), (path ? &lpath_0 : NULL), rep);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected $S [.'$R'.1]", (path ? &lpath_0 : NULL));
            }
        }
    }
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "not array or unexpected array [.'$R']", path);
    }
    return res;
}

static bool _jm_re_0(const char *s, jm_path_t *path, jm_report_t *rep)
{
    size_t slen = strlen(s);
    return cre2_match(_jm_re_0_re2, s, slen, 0, slen, CRE2_UNANCHORED, NULL, 0);
}

// check $S (.'$S')
static bool json_model_3(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$S'
    // "/[a-z]/"
    res = json_is_string(val) && _jm_re_0(json_string_value(val), path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected /[a-z]/ [.'$S']", path);
    }
    return res;
}

// check $ (.)
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .
    // .'|'.0
    res = json_model_3(val, path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected $S [.'|'.0]", path);
    }
    if (! res)
    {
        // .'|'.1
        res = json_model_5(val, path, rep);
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "unexpected $R#S [.'|'.1]", path);
        }
    }
    if (res)
    {
        if (rep) jm_report_free_entries(rep);
    }
    else
    {
        if (rep) jm_report_add_entry(rep, "no model matched [.'|']", path);
    }
    return res;
}

static bool _jm_re_1(const char *s, jm_path_t *path, jm_report_t *rep)
{
    size_t slen = strlen(s);
    return cre2_match(_jm_re_1_re2, s, slen, 0, slen, CRE2_UNANCHORED, NULL, 0);
}

// check $R#S (.'$R#S')
static bool json_model_5(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$R#S'
    // "/[0-9]/"
    res = json_is_string(val) && _jm_re_1(json_string_value(val), path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected /[0-9]/ [.'$R#S']", path);
    }
    return res;
}

jm_check_fun_t check_model_map(const char *pname)
{
    return jm_search_propmap(pname, check_model_map_tab, 3);
}

static bool initialized = false;

const char *check_model_init(void)
{
    if (!initialized)
    {
        initialized = true;
        jm_version_string = JSON_MODEL_VERSION;
        _jm_re_0_re2 = cre2_new("[a-z]", strlen("[a-z]"), NULL);
        if (cre2_error_code(_jm_re_0_re2))
            return cre2_error_string(_jm_re_0_re2);
        _jm_re_0_nn = cre2_num_capturing_groups(_jm_re_0_re2) + 1;
        _jm_re_1_re2 = cre2_new("[0-9]", strlen("[0-9]"), NULL);
        if (cre2_error_code(_jm_re_1_re2))
            return cre2_error_string(_jm_re_1_re2);
        _jm_re_1_nn = cre2_num_capturing_groups(_jm_re_1_re2) + 1;
        check_model_map_tab[0] = (jm_propmap_t) { "", json_model_1 };
        check_model_map_tab[1] = (jm_propmap_t) { "R", json_model_4 };
        check_model_map_tab[2] = (jm_propmap_t) { "S", json_model_3 };
        jm_sort_propmap(check_model_map_tab, 3);
    }
    return NULL;
}

void check_model_free(void)
{
    if (initialized)
    {
        initialized = false;

        // cleanup code
        cre2_delete(_jm_re_0_re2);
        _jm_re_0_re2 = NULL;
        _jm_re_0_nn = 0;
        cre2_delete(_jm_re_1_re2);
        _jm_re_1_re2 = NULL;
        _jm_re_1_nn = 0;
    }
}


/*
 * # API
 *
 * ```c
 * bool error;
 * char *message;
 *
 * bool valid = check(json_value, "model-name", &error, &message);
 *
 * if (message)
 *    free(message), message = NULL;
 * ```
 *
 * - `error` is set of there was an error during initialization or if the model does not exists.
 * - `message` is provided if a non NULL pointer is passed, and must be freed by the caller.
 */
bool
check_model(const json_t *val, const char *name, bool *error, char **reasons)
{
    return jm_generic_entry(check_model_init, check_model_map, val, name, error, reasons);
}
