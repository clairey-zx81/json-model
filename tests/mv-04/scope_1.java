//
// Generated by JSON Model Compiler version 2.0b1
// see https://github.com/clairey-zx81/json-model
//

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class scope_1 extends ModelChecker
{
    static public final String VERSION = "2";

    public Pattern _jm_re_0_pat = null;
    public Pattern _jm_re_1_pat = null;
    public Map<String, Checker> scope_1_map_pmap;

    // check $R (.'$R')
    public boolean json_model_4(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$R'
        res = json.isArray(val) && json.arrayLength(val) == 2;
        if (res)
        {
            Path lpath_0 = new Path(0, path);
            // .'$R'.0
            res = json_model_5(json.arrayItem(val, 0), (path != null ? lpath_0 : null), rep);
            if (! res)
            {
                if (rep != null) rep.addEntry("unexpected $S [.'$R'.0]", (path != null ? lpath_0 : null));
            }
            if (res)
            {
                lpath_0 = new Path(1, path);
                // .'$R'.1
                res = json_model_5(json.arrayItem(val, 1), (path != null ? lpath_0 : null), rep);
                if (! res)
                {
                    if (rep != null) rep.addEntry("unexpected $S [.'$R'.1]", (path != null ? lpath_0 : null));
                }
            }
        }
        if (! res)
        {
            if (rep != null) rep.addEntry("not array or unexpected array [.'$R']", path);
        }
        return res;
    }
    
    public boolean _jm_re_0(String val, Path path, Report rep)
    {
        return _jm_re_0_pat.matcher(val).find();
    }
    
    // check $S (.'$S')
    public boolean json_model_3(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$S'
        // "/[a-z]/"
        res = json.isString(val) && _jm_re_0(json.asString(val), path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected /[a-z]/ [.'$S']", path);
        }
        return res;
    }
    
    // check $ (.)
    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res;
        // a string with a lower case later or a digit
        // .
        // .'|'.0
        res = json_model_3(val, path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected $S [.'|'.0]", path);
        }
        if (! res)
        {
            // .'|'.1
            res = json_model_5(val, path, rep);
            if (! res)
            {
                if (rep != null) rep.addEntry("unexpected $R#S [.'|'.1]", path);
            }
        }
        if (res)
        {
            if (rep != null) rep.clearEntries();
        }
        else
        {
            if (rep != null) rep.addEntry("no model matched [.'|']", path);
        }
        return res;
    }
    
    public boolean _jm_re_1(String val, Path path, Report rep)
    {
        return _jm_re_1_pat.matcher(val).find();
    }
    
    // check $R#S (.'$R#S')
    public boolean json_model_5(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$R#S'
        // "/[0-9]/"
        res = json.isString(val) && _jm_re_1(json.asString(val), path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected /[0-9]/ [.'$R#S']", path);
        }
        return res;
    }
    

    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            _jm_re_0_pat = Pattern.compile("[a-z]");
            _jm_re_1_pat = Pattern.compile("[0-9]");
            scope_1_map_pmap = new HashMap<String, Checker>();
            scope_1_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
            scope_1_map_pmap.put("R", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_4(o, p, r);} });
            scope_1_map_pmap.put("S", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_3(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            _jm_re_0_pat = null;
            _jm_re_1_pat = null;
            scope_1_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return scope_1_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return scope_1_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new scope_1();
        Main.main("scope_1", checker, VERSION, args);
    }
}
