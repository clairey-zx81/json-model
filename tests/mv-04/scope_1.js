#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
//
// Generated by JSON Model Compiler version 2
// see https://github.com/clairey-zx81/json-model
//
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

const _jm_re_0_re = new runtime.RX("[a-z]", "")
const _jm_re_1_re = new runtime.RX("[0-9]", "")
var check_model_map = new Map()

// check $r (.'$r')
function json_model_4(val, path, rep)
{
    let res;
    // .'$r'
    res = Array.isArray(val) && val.length == 2;
    if (res)
    {
        let lpath_0 = path ? path.concat([0]) : null;
        // .'$r'.0
        res = json_model_5(val[0], (path ? lpath_0 : null), rep);
        if (! res)
        {
            rep !== null && rep.push(["unexpected $s [.'$r'.0]", (path ? lpath_0 : null)])
        }
        if (res)
        {
            lpath_0 = path ? path.concat([1]) : null;
            // .'$r'.1
            res = json_model_5(val[1], (path ? lpath_0 : null), rep);
            if (! res)
            {
                rep !== null && rep.push(["unexpected $s [.'$r'.1]", (path ? lpath_0 : null)])
            }
        }
    }
    if (! res)
    {
        rep !== null && rep.push(["not array or unexpected array [.'$r']", path])
    }
    return res;
}

const _jm_re_0 = (s) => _jm_re_0_re.exec(s) !== null

// check $s (.'$s')
function json_model_3(val, path, rep)
{
    let res;
    // .'$s'
    // "/[a-z]/"
    res = ((typeof val === 'string' || val instanceof String)) && _jm_re_0(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected /[a-z]/ [.'$s']", path])
    }
    return res;
}

// check $ (.)
function json_model_1(val, path, rep)
{
    let res;
    // a string with a lower case later or a digit
    // .
    // .'|'.0
    res = json_model_3(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected $s [.'|'.0]", path])
    }
    if (! res)
    {
        // .'|'.1
        res = json_model_5(val, path, rep);
        if (! res)
        {
            rep !== null && rep.push(["unexpected $r#s [.'|'.1]", path])
        }
    }
    if (res)
    {
        if (rep !== null) rep.length = 0
    }
    else
    {
        rep !== null && rep.push(["no model matched [.'|']", path])
    }
    return res;
}

const _jm_re_1 = (s) => _jm_re_1_re.exec(s) !== null

// check $r#s (.'$r#s')
function json_model_5(val, path, rep)
{
    let res;
    // .'$r#s'
    // "/[0-9]/"
    res = ((typeof val === 'string' || val instanceof String)) && _jm_re_1(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected /[0-9]/ [.'$r#s']", path])
    }
    return res;
}


var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        check_model_map.set("", json_model_1)
        check_model_map.set("r", json_model_4)
        check_model_map.set("s", json_model_3)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
    {
        initialized = false;
    }
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)
