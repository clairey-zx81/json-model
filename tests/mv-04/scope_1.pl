#! /usr/bin/perl
#
# Generated by JSON Model Compiler version 2
# see https://github.com/clairey-zx81/json-model
#
use strict;
use warnings;
use re::engine::RE2;
use JSON::JsonModel;
use constant JMC_VERSION => '2';


sub json_model_4($$$);
sub json_model_3($$$);
sub json_model_1($$$);
sub json_model_5($$$);
my %check_model_map;

# check $r (.'$r')
sub json_model_4($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .'$r'
    $res = jm_is_array($val) && scalar @$val == 2;
    if ($res)
    {
        # .'$r'.0
        $res = json_model_5($$val[0], undef, $rep);
        if ($res)
        {
            # .'$r'.1
            $res = json_model_5($$val[1], undef, $rep);
        }
    }
    return $res;
}

sub _jm_re_0($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = $val =~ /[a-z]/;
    return $res;
}

# check $s (.'$s')
sub json_model_3($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .'$s'
    # "/[a-z]/"
    $res = jm_is_string($val) && _jm_re_0($val, $path, $rep);
    return $res;
}

# check $ (.)
sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # a string with a lower case later or a digit
    # .
    # .'|'.0
    $res = json_model_3($val, $path, $rep);
    if (! $res)
    {
        # .'|'.1
        $res = json_model_5($val, $path, $rep);
    }
    return $res;
}

sub _jm_re_1($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = $val =~ /[0-9]/;
    return $res;
}

# check $r#s (.'$r#s')
sub json_model_5($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .'$r#s'
    # "/[0-9]/"
    $res = jm_is_string($val) && _jm_re_1($val, $path, $rep);
    return $res;
}


# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %check_model_map = (
            '' => \&json_model_1,
            'r' => \&json_model_4,
            's' => \&json_model_3,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %check_model_map = ();
    }
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();
