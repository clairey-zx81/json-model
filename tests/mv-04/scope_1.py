#! /bin/env python
#
# Generated by JSON Model Compiler version 2.0b0
# see https://github.com/clairey-zx81/json-model
#
from typing import Callable
import re2 as re
from json_model.runtime import *
__version__ = "2.0b0"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity agains JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

_jm_re_0_reco: object
_jm_re_0: RegexFun
_jm_re_1_reco: object
_jm_re_1: RegexFun
check_model_map: PropMap

# check $R (.'$R')
def json_model_4(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .'$R'
    res = isinstance(val, list) and len(val) == 2
    if res:
        lpath_0: Path = (path + [ 0 ]) if path is not None else None
        # .'$R'.0
        res = json_model_5(val[0], lpath_0 if path is not None else None, rep)
        if not res:
            rep is None or rep.append(("unexpected $S [.'$R'.0]", lpath_0 if path is not None else None))
        if res:
            lpath_0: Path = (path + [ 1 ]) if path is not None else None
            # .'$R'.1
            res = json_model_5(val[1], lpath_0 if path is not None else None, rep)
            if not res:
                rep is None or rep.append(("unexpected $S [.'$R'.1]", lpath_0 if path is not None else None))
    if not res:
        rep is None or rep.append(("not array or unexpected array [.'$R']", path))
    return res


# check $S (.'$S')
def json_model_3(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .'$S'
    # "/[a-z]/"
    res = isinstance(val, str) and _jm_re_0(val, path, rep)
    if not res:
        rep is None or rep.append(("unexpected /[a-z]/ [.'$S']", path))
    return res

# check $ (.)
def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # a string with a lower case later or a digit
    # .
    # .'|'.0
    res = json_model_3(val, path, rep)
    if not res:
        rep is None or rep.append(("unexpected $S [.'|'.0]", path))
    if not res:
        # .'|'.1
        res = json_model_5(val, path, rep)
        if not res:
            rep is None or rep.append(("unexpected $R#S [.'|'.1]", path))
    if res:
        rep is None or rep.clear()
    else:
        rep is None or rep.append(("no model matched [.'|']", path))
    return res


# check $R#S (.'$R#S')
def json_model_5(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .'$R#S'
    # "/[0-9]/"
    res = isinstance(val, str) and _jm_re_1(val, path, rep)
    if not res:
        rep is None or rep.append(("unexpected /[0-9]/ [.'$R#S']", path))
    return res


# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global _jm_re_0_reco, _jm_re_0
        _jm_re_0_reco = re.compile("[a-z]")
        _jm_re_0 = lambda s, p, r: _jm_re_0_reco.search(s) is not None
        global _jm_re_1_reco, _jm_re_1
        _jm_re_1_reco = re.compile("[0-9]")
        _jm_re_1 = lambda s, p, r: _jm_re_1_reco.search(s) is not None
        global check_model_map
        check_model_map = {
            "": json_model_1,
            "R": json_model_4,
            "S": json_model_3,
        }

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False
        global _jm_re_0_reco, _jm_re_0
        _jm_re_0_reco = None
        _jm_re_0 = None
        global _jm_re_1_reco, _jm_re_1
        _jm_re_1_reco = None
        _jm_re_1 = None

if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()
