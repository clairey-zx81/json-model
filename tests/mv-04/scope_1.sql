--
-- Generated by JSON Model Compiler version 2
-- see https://github.com/clairey-zx81/json-model
--
-- JSON_MODEL_VERSION is 2
CREATE EXTENSION IF NOT EXISTS json_model;

-- check $r (.'$r')
CREATE OR REPLACE FUNCTION json_model_4(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res bool;
BEGIN
  -- .'$r'
  res := JSONB_TYPEOF(val) = 'array' AND JSONB_ARRAY_LENGTH(val) = 2;
  IF res THEN
    -- .'$r'.0
    res := json_model_5(val -> 0, NULL, NULL);
    IF res THEN
      -- .'$r'.1
      res := json_model_5(val -> 1, NULL, NULL);
    END IF;
  END IF;
  RETURN res;
END;
$$ LANGUAGE PLpgSQL;

-- regex=[a-z] opts=n
CREATE OR REPLACE FUNCTION _jm_re_0(val TEXT, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
  RETURN regexp_like(val, '[a-z]', 'n');
END;
$$ LANGUAGE plpgsql;

-- check $s (.'$s')
CREATE OR REPLACE FUNCTION json_model_3(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
  -- .'$s'
  -- "/[a-z]/"
  RETURN JSONB_TYPEOF(val) = 'string' AND _jm_re_0(JSON_VALUE(val, '$' RETURNING TEXT), NULL, NULL);
END;
$$ LANGUAGE PLpgSQL;

-- check $ (.)
CREATE OR REPLACE FUNCTION json_model_1(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
  -- a string with a lower case later or a digit
  -- .
  -- .'|'.0
  -- .'|'.1
  RETURN json_model_3(val, NULL, NULL) OR json_model_5(val, NULL, NULL);
END;
$$ LANGUAGE PLpgSQL;

-- regex=[0-9] opts=n
CREATE OR REPLACE FUNCTION _jm_re_1(val TEXT, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
  RETURN regexp_like(val, '[0-9]', 'n');
END;
$$ LANGUAGE plpgsql;

-- check $r#s (.'$r#s')
CREATE OR REPLACE FUNCTION json_model_5(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
BEGIN
  -- .'$r#s'
  -- "/[0-9]/"
  RETURN JSONB_TYPEOF(val) = 'string' AND _jm_re_1(JSON_VALUE(val, '$' RETURNING TEXT), NULL, NULL);
END;
$$ LANGUAGE PLpgSQL;

CREATE OR REPLACE FUNCTION check_model_map(name TEXT)
RETURNS TEXT STRICT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  map JSONB := JSONB '{"":"json_model_1","r":"json_model_4","s":"json_model_3"}';
BEGIN
  RETURN map->>name;
END;
$$ LANGUAGE plpgsql;

--
-- constant maps initialization
--
TRUNCATE jm_constant_maps;

--
-- JSON Model checking entry point
--
-- TODO INOUT rep?
CREATE OR REPLACE FUNCTION check_model(val JSONB, name TEXT, rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  fun TEXT;
BEGIN
  fun := check_model_map(name);
  IF fun IS NULL THEN
    RAISE EXCEPTION 'model for % not found', name;
  END IF;
  RETURN jm_call(fun, val, NULL, rep);
END;
$$ LANGUAGE plpgsql;
