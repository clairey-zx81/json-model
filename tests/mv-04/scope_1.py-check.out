scope_1.values.json[0]: PASS
scope_1.values.json[1]: PASS
scope_1.values.json[2]: PASS
scope_1.values.json[3]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[4]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[5]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[6]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[7]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[8]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[9]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
scope_1.values.json[10]: FAIL [("unexpected REGEX [$.'$S']", []), ("unexpected $S [$.'|'.0]", []), ("unexpected REGEX [$.'$R#S']", []), ("unexpected $R#S [$.'|'.1]", []), ("no model matched [$.'|']", [])]
